### PR TITLE
dependency-track: init at 4.11.6

### PR DESCRIPTION
**Pull request by xanderio**
## Description of changes

https://dependencytrack.org/

I had to extend `buildMavenPackage` with `afterDepsSetup`, this is a string of build phase commands that are injected after the `nvmDeps` variable is set up. This is needed as dependency-track builds its own BOM during its build and expects this file to be presented in a later build step.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by xanderio** [comment]
The module is veeery basic at the moment and is still missing proper database support amount other things. I should hopefully get this sorted later today or tomorrow.

File: nixos/tests/all-tests.nix [hunk 1]
```diff
@@ -250,6 +250,7 @@ in {
   deepin = handleTest ./deepin.nix {};
   deluge = handleTest ./deluge.nix {};
   dendrite = handleTest ./matrix/dendrite.nix {};
+  dependency-track = handleTest ./dependency-track.nix {};
```
Comment (AndersonTorres) [inline nixos/tests/all-tests.nix:253]: Split the test module in its own commit.

File: pkgs/by-name/de/dependency-track/package.nix [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    owner = "DependencyTrack";
+    repo = "dependency-track";
+    rev = version;
+    hash = "sha256-iVJidl1NbgTj5+DiU8MKJy0lMo608cfA2oMcrypLadI=";
+  };
+
+  patches = [
+    ./remove-frontend-download.patch
+    ./add-junixsocket.patch
+  ];
```
Comment (AndersonTorres) [inline pkgs/by-name/de/dependency-track/package.nix:30]: Prefix the patches with numbers, so that they are better viewed in the file manager and GitHub:

```suggestion
  patches = [
    ./0000-remove-frontend-download.patch
    ./0001-add-junixsocket.patch
  ];
```

File: pkgs/by-name/de/dependency-track/package.nix [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    runHook preInstall
+
+    install -Dm644 target/dependency-track-*.jar $out/share/dependency-track/dependency-track.jar
+    makeWrapper ${jre_headless}/bin/java $out/bin/dependency-track \
+      --add-flags "-jar $out/share/dependency-track/dependency-track.jar"
+
+    runHook postInstall
+  '';
+
+  frontend = buildNpmPackage {
```
Comment (AndersonTorres) [inline pkgs/by-name/de/dependency-track/package.nix:76]: Move `frontend`  to let-in block.

File: pkgs/by-name/de/dependency-track/package.nix [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  };
+
+  passthru = {
+    tests = {
+      inherit (nixosTests) dependency-track;
+    };
+    updateScript = nix-update-script { };
+  };
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/by-name/de/dependency-track/package.nix:102]:
```suggestion
  meta = {
```

File: pkgs/by-name/de/dependency-track/package.nix [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    description = "Dependency-Track is an intelligent Component Analysis platform that allows organizations to identify and reduce risk in the software supply chain";
```
Comment (AndersonTorres) [inline pkgs/by-name/de/dependency-track/package.nix:103]:
```suggestion
    description = "Intelligent Component Analysis platform that allows organizations to identify and reduce risk in the software supply chain";
```

File: pkgs/by-name/de/dependency-track/package.nix [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    homepage = "https://github.com/DependencyTrack/dependency-track";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ xanderio ];
```
Comment (AndersonTorres) [inline pkgs/by-name/de/dependency-track/package.nix:106]:
```suggestion
    license = lib.licenses.asl20;
    maintainers = with lib.maintainers; [ xanderio ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/de/dependency-track/package.nix [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    npmDepsHash = "sha256-LHK222OzjVHYP4/cVpWfoFSFbfbEt5FNTJry5te3XvQ=";
+    forceGitDeps = true;
+    makeCacheWritable = true;
+
+    # The prepack script runs the build script, which we'd rather do in the build phase.
+    npmPackFlags = [ "--ignore-scripts" ];
+  };
+in
+
+maven.buildMavenPackage rec {
```
Comment (AndersonTorres) [inline pkgs/by-name/de/dependency-track/package.nix:35]: I believe this `rec` is not needed.

**Review by AndersonTorres — COMMENTED** [review]
Result of `nixpkgs-review pr 328669` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>33 packages built:</summary>
  <ul>
    <li>apgdiff</li>
    <li>bytecode-viewer</li>
    <li>certificate-ripper</li>
    <li>commafeed</li>
    <li>cryptomator</li>
    <li>dependency-track</li>
    <li>digital</li>
    <li>forge-mtg</li>
    <li>gephi</li>
    <li>global-platform-pro</li>
    <li>gol</li>
    <li>h2</li>
    <li>h2.doc</li>
    <li>java-language-server</li>
    <li>jd-cli</li>
    <li>joularjx</li>
    <li>jugglinglab</li>
    <li>kotlin-interactive-shell</li>
    <li>ktfmt</li>
    <li>lemminx</li>
    <li>mariadb-connector-java</li>
    <li>ns-usbloader</li>
    <li>openrefine</li>
    <li>protege</li>
    <li>quark-goldleaf</li>
    <li>runelite</li>
    <li>s3proxy</li>
    <li>scenebuilder</li>
    <li>schemaspy</li>
    <li>slipstream</li>
    <li>sonarlint-ls</li>
    <li>tabula-java</li>
    <li>tika</li>
  </ul>
</details>

Why so many packages are being built?

**Comment by xanderio** [comment]
Unfortunately there is a bit more work required to support configuring the frontend for OIDC support. This will require us to provide a nginx virtualHost configuration.

**Comment by xanderio** [comment]
> Result of `nixpkgs-review pr 328669` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 2 packages blacklisted:
> 33 packages built:
> 
> Why so many packages are being built?

Do to the change in `buildMavenPackage` as stated above.

**Comment by xanderio** [comment]
> Unfortunately there is a bit more work required to support configuring the frontend for OIDC support. This will require us to provide a nginx virtualHost configuration.

Done

File: pkgs/by-name/de/dependency-track/package.nix (hunk 7 above)
Comment (xanderio) [reply pkgs/by-name/de/dependency-track/package.nix:35]: It's needed for the `src` reference in line 32.

File: pkgs/by-name/ma/maven/build-maven-package.nix [hunk 8]
```diff
@@ -93,6 +94,7 @@ stdenv.mkDerivation (builtins.removeAttrs args [ "mvnFetchExtraArgs" ] // {
     runHook preBuild
 
     mvnDeps=$(cp -dpR ${fetchedMavenDeps}/.m2 ./ && chmod +w -R .m2 && pwd)
+    ${afterDepsSetup}
```
Comment (AndersonTorres) [inline pkgs/by-name/ma/maven/build-maven-package.nix:97]: where and why was this added?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ma/maven/build-maven-package.nix (hunk 8 above)
Comment (xanderio) [reply pkgs/by-name/ma/maven/build-maven-package.nix:97]: As stated above:

> I had to extend `buildMavenPackage` with `afterDepsSetup`, this is a string of build phase commands that are injected after the nvmDeps variable is set up. This is needed as dependency-track builds its own BOM during its build and expects this file to be presented in a later build step.

Comment (AndersonTorres) [reply pkgs/by-name/ma/maven/build-maven-package.nix:97]: Since it changes maven builder - a thing that will affect all packages that use Maven - it should be in a separate **pull request**, so that it can be bisect-able in the future.

Comment (leona-ya) [reply pkgs/by-name/ma/maven/build-maven-package.nix:97]: Why should it be a seperate PR? A seperate commit should be bisectable as well.

Comment (AndersonTorres) [reply pkgs/by-name/ma/maven/build-maven-package.nix:97]: Mostly because GitHub is a mess :(
GitHub is PR- and merge-centric.

Because of it, it is better to at least open a PR for the change in Maven, so that the discussions about Maven (that affects all Java+Maven packages) can happen without clobbering the discussion on the inclusion of this package.

This is more preeminent given that when I tried to run nixpkgs-review it got stuck building many unrelated packages.

(this is why so many oldschoool developers use e-mail...)

Comment (AndersonTorres) [reply pkgs/by-name/ma/maven/build-maven-package.nix:97]: Another thing: GitHub allows comments on commits, but only after they have been committed.

File: nixos/modules/services/web-apps/dependency-track.nix [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+        User = "postgres";
+        Group = "postgres";
+        LoadCredential = [ "db_password:${cfg.database.passwordFile}" ];
+      };
+      script = ''
+        set -o errexit -o pipefail -o nounset -o errtrace
+        shopt -s inherit_errexit
+
+        create_role="$(mktemp)"
+        trap 'rm -f "$create_role"' EXIT
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/dependency-track.nix:531]: Please use PrivateTmp instead

File: nixos/modules/services/web-apps/dependency-track.nix [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+        db_password="$(<"$CREDENTIALS_DIRECTORY/db_password")"
+        db_password="''${db_password//\'/\'\'}"
+
+        echo "CREATE ROLE \"dependency-track\" WITH LOGIN PASSWORD '$db_password' CREATEDB" > "$create_role"
+        psql -tAc "SELECT 1 FROM pg_roles WHERE rolname='dependency-track'" | grep -q 1 || psql -tA --file="$create_role"
+        psql -tAc "SELECT 1 FROM pg_database WHERE datname = 'dependency-track'" | grep -q 1 || psql -tAc 'CREATE DATABASE "dependency-track" OWNER "dependency-track"'
+      '';
+    };
+
+    services.postgresql.enable = lib.mkDefault cfg.database.createLocally;
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/dependency-track.nix:546]:
```suggestion
    services.postgresql.enable = lib.mkIf cfg.database.createLocally (lib.mkDefault true);
```

File: nixos/modules/services/web-apps/dependency-track.nix [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+          Group = "dependency-track";
+          DynamicUser = true;
+          StateDirectory = "dependency-track";
+          LoadCredential =
+            [ "db_password:${cfg.database.passwordFile}" ]
+            ++ lib.optional cfg.settings."alpine.ldap.enabled"
+              "ldap_bind_password:${cfg.ldap.bindPasswordFile}";
+        };
+        script = ''
+          set -o errexit -o pipefail -o nounset -o errtrace
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/dependency-track.nix:575]:
```suggestion
          set -eou pipefail
```

File: nixos/modules/services/web-apps/dependency-track.nix [hunk 12]
```diff
@@ -0,0 +1,602 @@
...
+    if lib.isPath value then
+      throw ''
+        services.dependency-track.${optionName}:
+          ${toString value}
+          is a Nix path, but should be a string, since Nix
+          paths are copied into the world-readable Nix store.
+      ''
+    else
+      value;
+
```
[the remark this answers is not in the document]
Comment (xanderio) [reply nixos/modules/services/web-apps/dependency-track.nix:39]: It doesn't break agenix, so I would assume this also doesn't break sops-nix.

File: nixos/tests/dependency-track.nix [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+                touch "$NIST_DIR/nvdcve-1.1-$feed.json.gz"
+                echo "9999999999999" > "$NIST_DIR/nvdcve-1.1-$feed.json.gz.ts"
+              done
+            '';
+          };
+          services.dependency-track = {
+            enable = true;
+            port = dependencyTrackPort;
+            nginx.domain = "localhost";
+            database.passwordFile = "${pkgs.writeText "dbPassword" ''wzf6\"vO"Cb\nP>p#6;c&o?eu=q'THE'''H''''E''}";
```
[the remark this answers is not in the document]
Comment (xanderio) [reply nixos/tests/dependency-track.nix]: Change the majority to hunter2, the `'THE'''H''''E` is testing for the escapement in the Postgres setup unit.

File: pkgs/by-name/ma/maven/build-maven-package.nix (hunk 8 above)
[the remark this answers is not in the document]
Comment (xanderio) [reply pkgs/by-name/ma/maven/build-maven-package.nix]: I've change this to use `runHook`.

**Comment by xanderio** [comment]
I'm currently debugging a OIDC team synchronization issue.

**Comment by xanderio** [comment]
I have found the issue and fix it. We are again ready to merge from my end.

File: nixos/modules/services/web-apps/dependency-track.nix [hunk 14]
```diff
@@ -0,0 +1,604 @@
...
+        databaseServices = [
+          "dependency-track-postgresql-init.service"
+          "postgresql.service"
+        ];
+      in
+      {
+        description = "Dependency Track";
+        wantedBy = [ "multi-user.target" ];
+        requires = databaseServices;
+        after = databaseServices;
```
Comment (leona-ya) [inline nixos/modules/services/web-apps/dependency-track.nix:561]: This will probably break when no postgresql exists on the system.

**Review by leona-ya — APPROVED** [review]

**Review by leona-ya — APPROVED** [review]

**Comment by xanderio** [comment]
@SuperSandro2000 is there anything left to-do before this can be merged? The long time to merge is quite demotivating. :)

**Comment by leona-ya** [comment]
It seems the frontend source hash changed :/

```
error: hash mismatch in fixed-output derivation '/nix/store/qnmrsbx2zipr5zc7b86c33agfk55x2qd-source.drv':
        likely URL: https://github.com/DependencyTrack/frontend/archive/4.11.6.tar.gz
         specified: sha256-YJ5D7hTLlaAgOXz4nT+eh1n5/OulHCsPCN1LAB+1J8M=
            got:    sha256-RTOJXrdJZqY6/zTMY7eBGXb5rEoKos78ZGVQhE3qIFg=
error: 1 dependencies of derivation '/nix/store/yjdr7lp5l38lms5jzk1jsppd60za07h7-dependency-track-frontend-4.11.6-npm-deps.drv' failed to build

**Review by leona-ya — APPROVED** [review]
